source/mqtt3_to_mqtt5_adapter.c

```diff
@@ -260,8 +269,451 @@ static int s_aws_mqtt5_adapter_perform_safe_callback(
+}
+
+static void s_adapter_connect_task_fn(struct aws_task *task, void *arg, enum aws_task_status status);
+
+static struct aws_mqtt_adapter_connect_task *s_aws_mqtt_adapter_connect_task_new(
+    struct aws_allocator *allocator,
+    struct aws_mqtt_client_connection_5_impl *adapter,
+    const struct aws_mqtt_connection_options *connection_options) {
+
+    struct aws_mqtt_adapter_connect_task *connect_task =
+        aws_mem_calloc(allocator, 1, sizeof(struct aws_mqtt_adapter_connect_task));
+
+    aws_task_init(&connect_task->task, s_adapter_connect_task_fn, (void *)connect_task, "AdapterConnectTask");
+    connect_task->allocator = adapter->allocator;
+    connect_task->adapter = (struct aws_mqtt_client_connection_5_impl *)aws_ref_count_acquire(&adapter->internal_refs);
+
+    aws_byte_buf_init_copy_from_cursor(&connect_task->host_name, allocator, connection_options->host_name);
+    connect_task->port = connection_options->port;
+    connect_task->socket_options = *connection_options->socket_options;
+    if (connection_options->tls_options) {
+        aws_tls_connection_options_copy(&connect_task->tls_options, connection_options->tls_options);
+        connect_task->tls_options_ptr = &connect_task->tls_options;
+    }
+
+    aws_byte_buf_init_copy_from_cursor(&connect_task->client_id, allocator, connection_options->client_id);
+
+    connect_task->keep_alive_time_secs = connection_options->keep_alive_time_secs;
+    connect_task->ping_timeout_ms = connection_options->ping_timeout_ms;
+    connect_task->protocol_operation_timeout_ms = connection_options->protocol_operation_timeout_ms;
+    connect_task->on_connection_complete = connection_options->on_connection_complete;
+    connect_task->on_connection_complete_user_data = connection_options->user_data;
+    connect_task->clean_session = connection_options->clean_session;
+
+    return connect_task;
+}
+
+static int s_aws_mqtt_client_connection_5_connect(
+    void *impl,
+    const struct aws_mqtt_connection_options *connection_options) {
+
+    struct aws_mqtt_client_connection_5_impl *adapter = impl;
+
+    struct aws_mqtt_adapter_connect_task *task =
+        s_aws_mqtt_adapter_connect_task_new(adapter->allocator, adapter, connection_options);
+    if (task == NULL) {
+        AWS_LOGF_ERROR(AWS_LS_MQTT_CLIENT, "id=%p: failed to create adapter connect task", (void *)adapter);
+        return AWS_OP_ERR;
+    }
+
+    aws_event_loop_schedule_task_now(adapter->loop, &task->task);
+
+    return AWS_OP_SUCCESS;
+}
+
+static int s_aws_mqtt3_to_mqtt5_adapter_safe_lifecycle_handler(
+    struct aws_mqtt_client_connection_5_impl *adapter,
+    void *context) {
+    const struct aws_mqtt5_client_lifecycle_event *event = context;
+
+    /*
+     * Never invoke a callback after termination
+     */
+    if (adapter->synced_data.terminated) {
+        return AWS_OP_SUCCESS;
+    }
+
+    switch (event->event_type) {
+
+        case AWS_MQTT5_CLET_CONNECTION_SUCCESS:
```

Comment:
Just noting the obvious that we'll need to add the new mqtt3 callbacks into this section once they're implemented elsewhere -_-;;

Reply:
Ah, we do this validation in the aws_mqtt5_client_options_validate() but I think this implementation currently bypasses that validation and directly sets the stored config. Should we process these options through validation or just allow things to potentially fail on mqtt5 attempting to connect with bad settings?

Reply:
Added keep alive validation for now, can consider other fields as well, but doesn't seem pressing.